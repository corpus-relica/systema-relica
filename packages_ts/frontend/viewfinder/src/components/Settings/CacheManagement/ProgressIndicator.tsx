import React from 'react';

interface ProgressIndicatorProps {
  progress: number;
  phase: string;
}

const ProgressIndicator = ({
  progress,
  phase
}: ProgressIndicatorProps) => {
  return (
    <div className="space-y-2">
      <div className="flex justify-between text-sm">
        <span>{phase || 'Rebuilding cache...'}</span>
        <span>{Math.round(progress)}%</span>
      </div>
      <div className="w-full bg-gray-200 rounded-full h-2.5">
        <div
          className="bg-blue-600 h-2.5 rounded-full transition-all duration-300"
          style={{ width: `${progress}%` }}
        />
      </div>
    </div>
  );
};

export default ProgressIndicator;