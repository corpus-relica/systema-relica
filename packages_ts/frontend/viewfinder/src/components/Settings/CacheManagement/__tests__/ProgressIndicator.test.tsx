import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import ProgressIndicator from '../ProgressIndicator';

describe('ProgressIndicator', () => {
  it('renders progress bar with correct percentage', () => {
    render(<ProgressIndicator progress={45} phase="Processing entities" />);
    
    expect(screen.getByText('45%')).toBeInTheDocument();
    expect(screen.getByText('Processing entities')).toBeInTheDocument();
  });

  it('renders progress bar at 0%', () => {
    render(<ProgressIndicator progress={0} phase="Initializing" />);
    
    expect(screen.getByText('0%')).toBeInTheDocument();
    expect(screen.getByText('Initializing')).toBeInTheDocument();
  });

  it('renders progress bar at 100%', () => {
    render(<ProgressIndicator progress={100} phase="Completed" />);
    
    expect(screen.getByText('100%')).toBeInTheDocument();
    expect(screen.getByText('Completed')).toBeInTheDocument();
  });

  it('renders progress bar with correct width style', () => {
    const { container } = render(<ProgressIndicator progress={75} phase="Updating cache" />);
    
    const progressBar = container.querySelector('.bg-blue-600');
    expect(progressBar).toHaveStyle({ width: '75%' });
  });

  it('handles edge case of negative progress', () => {
    render(<ProgressIndicator progress={-10} phase="Error state" />);
    
    const progressBar = screen.getByRole('progressbar');
    expect(progressBar).toHaveAttribute('aria-valuenow', '0');
  });

  it('handles edge case of progress over 100', () => {
    render(<ProgressIndicator progress={150} phase="Overflow" />);
    
    const progressBar = screen.getByRole('progressbar');
    expect(progressBar).toHaveAttribute('aria-valuenow', '100');
  });

  it('renders with proper accessibility attributes', () => {
    render(<ProgressIndicator progress={50} phase="Processing" />);
    
    const progressBar = screen.getByRole('progressbar');
    expect(progressBar).toHaveAttribute('aria-valuemin', '0');
    expect(progressBar).toHaveAttribute('aria-valuemax', '100');
    expect(progressBar).toHaveAttribute('aria-valuenow', '50');
    expect(progressBar).toHaveAttribute('aria-label', 'Cache rebuild progress');
  });

  it('displays phase text correctly', () => {
    const phases = [
      'Initializing cache rebuild',
      'Processing entity cache',
      'Processing relationship cache',
      'Finalizing updates',
      'Completed successfully'
    ];
    
    phases.forEach(phase => {
      const { rerender } = render(<ProgressIndicator progress={50} phase={phase} />);
      expect(screen.getByText(phase)).toBeInTheDocument();
      rerender(<ProgressIndicator progress={50} phase="" />);
    });
  });

  it('takes a snapshot at different progress levels', () => {
    const progressLevels = [0, 25, 50, 75, 100];
    
    progressLevels.forEach(progress => {
      const { container } = render(
        <ProgressIndicator 
          progress={progress} 
          phase={`Processing at ${progress}%`} 
        />
      );
      expect(container).toMatchSnapshot(`progress-${progress}`);
    });
  });

  it('applies correct CSS classes for styling', () => {
    const { container } = render(<ProgressIndicator progress={60} phase="Processing" />);
    
    const wrapper = container.firstChild;
    expect(wrapper).toHaveClass('w-full');
    
    const progressContainer = container.querySelector('.bg-gray-200');
    expect(progressContainer).toHaveClass('rounded-full', 'h-2.5');
    
    const progressBar = container.querySelector('.bg-blue-600');
    expect(progressBar).toHaveClass('h-2.5', 'rounded-full', 'transition-all', 'duration-300');
  });

  it('renders empty phase gracefully', () => {
    render(<ProgressIndicator progress={30} phase="" />);
    
    expect(screen.getByText('30%')).toBeInTheDocument();
    // Should not throw error with empty phase
  });
});