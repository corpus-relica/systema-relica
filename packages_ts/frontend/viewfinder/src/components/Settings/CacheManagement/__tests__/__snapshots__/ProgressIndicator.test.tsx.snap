// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressIndicator takes a snapshot at different progress levels: progress-0 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <div
      class="flex justify-between text-sm"
    >
      <span>
        Processing at 0%
      </span>
      <span>
        0
        %
      </span>
    </div>
    <div
      class="w-full bg-gray-200 rounded-full h-2.5"
    >
      <div
        class="bg-blue-600 h-2.5 rounded-full transition-all duration-300"
        style="width: 0%;"
      />
    </div>
  </div>
</div>
`;

exports[`ProgressIndicator takes a snapshot at different progress levels: progress-25 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <div
      class="flex justify-between text-sm"
    >
      <span>
        Processing at 25%
      </span>
      <span>
        25
        %
      </span>
    </div>
    <div
      class="w-full bg-gray-200 rounded-full h-2.5"
    >
      <div
        class="bg-blue-600 h-2.5 rounded-full transition-all duration-300"
        style="width: 25%;"
      />
    </div>
  </div>
</div>
`;

exports[`ProgressIndicator takes a snapshot at different progress levels: progress-50 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <div
      class="flex justify-between text-sm"
    >
      <span>
        Processing at 50%
      </span>
      <span>
        50
        %
      </span>
    </div>
    <div
      class="w-full bg-gray-200 rounded-full h-2.5"
    >
      <div
        class="bg-blue-600 h-2.5 rounded-full transition-all duration-300"
        style="width: 50%;"
      />
    </div>
  </div>
</div>
`;

exports[`ProgressIndicator takes a snapshot at different progress levels: progress-75 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <div
      class="flex justify-between text-sm"
    >
      <span>
        Processing at 75%
      </span>
      <span>
        75
        %
      </span>
    </div>
    <div
      class="w-full bg-gray-200 rounded-full h-2.5"
    >
      <div
        class="bg-blue-600 h-2.5 rounded-full transition-all duration-300"
        style="width: 75%;"
      />
    </div>
  </div>
</div>
`;

exports[`ProgressIndicator takes a snapshot at different progress levels: progress-100 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <div
      class="flex justify-between text-sm"
    >
      <span>
        Processing at 100%
      </span>
      <span>
        100
        %
      </span>
    </div>
    <div
      class="w-full bg-gray-200 rounded-full h-2.5"
    >
      <div
        class="bg-blue-600 h-2.5 rounded-full transition-all duration-300"
        style="width: 100%;"
      />
    </div>
  </div>
</div>
`;
